src/VisualStudio/Core/Def/Implementation/ProjectSystem/RuleSets/VisualStudioRuleSetManager.cs

```diff
@@ -25,35 +40,54 @@ internal sealed partial class VisualStudioRuleSetManager : IWorkspaceService, ID
+            ReferenceCountedDisposable<RuleSetFile> disposable = null;
+
+            lock (_gate)
             {
-                ruleSetFile = new RuleSetFile(ruleSetFileFullPath, _fileChangeService, this);
-                _ruleSetFileMap.Add(ruleSetFileFullPath, ruleSetFile);
+                // If we already have one in the map to hand out, great
+                if (_ruleSetFileMap.TryGetValue(ruleSetFileFullPath, out var weakReference))
+                {
+                    disposable = weakReference.TryAddReference();
+                }
+
+                if (disposable == null)
+                {
+                    // We didn't easily get a disposable, so one of two things is the case:
+                    //
+                    // 1. We have no entry in _ruleSetFileMap at all for this.
+                    // 2. We had an entry, but it was disposed and is no longer valid.
+
+                    // In either case, we'll create a new rule set file and add it to the map.
+                    disposable = new ReferenceCountedDisposable<RuleSetFile>(new RuleSetFile(ruleSetFileFullPath, this));
+                    _ruleSetFileMap[ruleSetFileFullPath] = new ReferenceCountedDisposable<RuleSetFile>.WeakReference(disposable);
+                }
             }
 
-            return ruleSetFile;
-        }
+            // Call InitializeFileTracking outside the lock, so we don't have requests for other files blocking behind the initialization of this one.
+            // RuleSetFile itself will ensure InitializeFileTracking is locked as appropriate.
+            disposable.Target.InitializeFileTracking(_fileChangeService);
 
-        private void StopTrackingRuleSetFile(RuleSetFile ruleSetFile)
-        {
-            _ruleSetFileMap.Remove(ruleSetFile.FilePath);
+            return disposable;
         }
 
-        public void ClearCachedRuleSetFiles()
+        private void StopTrackingRuleSetFile(RuleSetFile ruleSetFile)
         {
-            foreach (var pair in _ruleSetFileMap)
+            // We can arrive here in one of two situations:
+            // 
+            // 1. The underlying RuleSetFile was disposed by all consumers, and we can try cleaning up our weak reference. This is purely an optimization
+            //    to avoid the key/value pair being unnecessarily held.
+            // 2. The RuleSetFile was modified, and we want to get rid of our cache now. Anybody still holding onto the values will dispose at their leaisure,
+            //    but it won't really matter anyways since the Dispose() that cleans up file trackers is already done.
+            //
+            // In either case, we can just be lazy and remove the key/value pair. It's entirely possible that we're seeing a file watcher event be raised _very_ lazily,
+            // and we've already created a new entry. But we have no good API for asking the ReferenceCountedDisposable.WeakReference we have in the map "are you definitely
```

Comment:
:bulb: Consider one of the following:

1. Use the following code to check instances

    ```csharp
    if (_ruleSetFileMap.TryGetValue(ruleSetFile, out var weakReference)) {
      using (var reference = weakReference.TryGetReference()) {
        if (reference == null || reference.Target == ruleSetFile) {
          _ruleSetFileMap.Remove(ruleSetFile.FilePath);
        }
      }
    }
    ```

2. Modify the comment to instead explain that clearing the cache for a different RuleSetFile is not problematic (correctness, plus the performance cost of recreating the item amortized against the likelihood that a mismatch occurs)

Reply:
I wanted to avoid 1 to keep it simple. Comment was intended to do 2 but I'll clarify further.